File: schemas/messaging-chat/drizzle/conversations.ts
// conversations: core conversation container for direct messages, group chats, and channels.
// See README.md for full design rationale.

import { pgTable, pgEnum, uuid, text, boolean, integer, timestamp, index } from "drizzle-orm/pg-core";
import { sql } from "drizzle-orm";
import { users } from "./users";

export const conversationTypeEnum = pgEnum("conversation_type", ["direct", "group", "channel"]);

export const conversations = pgTable("conversations", {
    id: uuid("id").primaryKey().defaultRandom(),
    type: conversationTypeEnum("type").notNull(),
    name: text("name"),
    description: text("description"),
    slug: text("slug").unique(),
    isPrivate: boolean("is_private").notNull().default(false),
    isArchived: boolean("is_archived").notNull().default(false),
    lastMessageAt: timestamp("last_message_at", { withTimezone: true }),
    messageCount: integer("message_count").notNull().default(0),
    createdBy: uuid("created_by").notNull().references(() => users.id, { onDelete: "restrict" }),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    index("idx_conversations_type").on(table.type),
    index("idx_conversations_is_private_type").on(table.isPrivate, table.type),
    index("idx_conversations_created_by").on(table.createdBy),
    index("idx_conversations_last_message_at").on(table.lastMessageAt),
  ]
);

File: schemas/messaging-chat/spacetimedb/conversations.rs
// conversations: chat conversations (direct, group, or channel) between users.
// See README.md for full design rationale.

use spacetimedb::{SpacetimeType, Timestamp};

#[derive(SpacetimeType, Clone)]
pub enum ConversationType {
    Direct, // type: String
    Group,
    Channel,
}

#[spacetimedb::table(name = conversations, public)]
pub struct Conversation {
    #[primary_key]
    pub id: String, // UUID
    #[index(btree)]
    pub conversation_type: ConversationType,
    pub name: Option<String>,
    pub description: Option<String>,
    #[unique]
    pub slug: Option<String>,
    // Composite index: (is_private, conversation_type) — not supported in SpacetimeDB; document only.
    #[index(btree)]
    pub is_private: bool,
    pub is_archived: bool,
    #[index(btree)]
    pub last_message_at: Option<Timestamp>,
    pub message_count: i32,
    #[index(btree)]
    pub created_by: String, // UUID — FK → users.id (restrict delete)
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
